### PR TITLE
Publish information about state snapshots to the network

### PR DESCRIPTION
**Pull request by saketh-are**
Adds a new cache SnapshotHosts which contains information about state snapshots hosted by peers in the network.  It will enable state part requests to be made directly to peers having the desired state.

Whenever a node produces a new snapshot, it will sign and publish information to the network specifying:
- Sync hash
- Epoch height
- The shards included in the snapshot

The published information is spread to all network nodes:
- Whenever a node receives new information, it broadcasts it to all of its neighbors
- Whenever a new connection is formed, the two nodes perform a full sync of available information

An RPC endpoint `/debug/api/snapshot_hosts` is provided for debugging purposes, as well as a display in the debug UI:

image

### CODE REVIEW

File: chain/chain/src/state_snapshot_actor.rs [hunk 1]
```diff
@@ -90,7 +100,20 @@ impl actix::Handler<WithSpanContext<CreateSnapshotRequest>> for StateSnapshotAct
             tracing::error!(target: "state_snapshot", ?prev_block_hash, ?shard_uids, "Failed to unlock flat state updates");
         }
         match res {
-            Ok(_) => {
+            Ok(res_shard_uids) => {
+                if let Some(res_shard_uids) = res_shard_uids {
+                    self.network_adapter.send(PeerManagerMessageRequest::NetworkRequests(
+                        NetworkRequests::SnapshotHostInfo {
```
Comment (saketh-are) [inline chain/chain/src/state_snapshot_actor.rs:106]: Reviewer(s): please double check the properties of the StateSnapshot recorded here.

Comment (nikurt) [reply chain/chain/src/state_snapshot_actor.rs:106]: Took me more than a moment to realize what's happening to the set of shard_uids. Looks good.

File: chain/network/src/peer_manager/connection/mod.rs [hunk 2]
```diff
@@ -182,6 +184,48 @@ impl Connection {
             }
         }
     }
+
+    pub fn send_snapshot_hosts(
```
Comment (nikurt) [inline chain/network/src/peer_manager/connection/mod.rs:188]: Can you please add some comments for this function?
What are `ds` and `sum`? Why does `sum` need to be computed here? Do you assume that `shard_ids` match exactly for the same peer/epoch_height ?

**Review by nikurt — APPROVED** [review]

File: chain/network/src/network_protocol/state_sync.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+pub struct SnapshotHostInfo {
+    /// Id of the node serving the snapshot
+    pub peer_id: PeerId,
+    /// Hash of the snapshot's state root
+    pub sync_hash: CryptoHash,
+    /// Ordinal of the epoch of the state root
+    pub epoch_height: EpochHeight,
+    /// List of shards included in the snapshot
+    pub shards: Vec<ShardId>,
+    /// Signature on (sync_hash, shards)
```
Comment (saketh-are) [inline chain/network/src/network_protocol/state_sync.rs:26]: This comment should say `(sync_hash, epoch_height, shards)`

**Pull request merged**